### DISCUSSION TITLE
performance optimization

### DISCUSSION THREAD

**Discussion by rentaohu** [ppiastucki/recast4j#149] [Ideas]
detour/src/main/java/org/recast4j/detour/NodePool.java#L73 
The get in this place seems redundant. Because this method is a high-frequency operation, there should be a lot of performance improvement when the list is passed in. 
detour/src/main/java/org/recast4j/detour/NavMeshQuery#getPathToNode   
Adding node here uses add(0, node), there will be a large number of array element copy operations, use add(node), and finally reverse, the performance should improve.

I really appreciate your work on the recast4j project, I was using the c++ version before and recently I wanted to try recast4j. Do you know if there is any running project using recast4j, or has recast4j done a lot of verification tests?
really appreciate your.

**Comment by ppiastucki** [comment]
Thanks for reporting the issues you spotted. I believe there may be many more places where the code is not optimal as during the initial porting I tried to keep the code as similar to the c++ version as possible even at the expense of performance. 
I will incorporate the changes you mentioned. Based on some preliminary tests they improve the performance by roughly 5% so do not expect a huge difference unless you work with particularly long paths.
As this is an open source project I do not know exactly who and how uses it. I know for instance that the library has been used quite extensively in https://atavismonline.com/ for a few years.
